Mathlib/NumberTheory/LSeries/ZMod.lean

```diff
@@ -0,0 +1,203 @@
+-/
+
+open HurwitzZeta Complex ZMod Finset Classical Topology Filter
+
+open scoped Real
+
+namespace ZMod
+
+variable {N : ℕ} [NeZero N]
+
+/-- If `Φ` is a periodic function, then the L-series of `Φ` converges for `1 < re s`. -/
+lemma LSeriesSummable_of_one_lt_re (Φ : ZMod N → ℂ) {s : ℂ} (hs : 1 < re s) :
+    LSeriesSummable (Φ ·) s := by
+  let c := max' _ <| univ_nonempty.image (Complex.abs ∘ Φ)
+  refine LSeriesSummable_of_bounded_of_one_lt_re (fun n _ ↦ le_max' _ _ ?_) (m := c) hs
+  exact mem_image_of_mem _ (mem_univ _)
+
+/--
+The unique meromorphic function `ℂ → ℂ` which agrees with `∑' n : ℕ, Φ n / n ^ s` wherever the
+latter is convergent. This is constructed as a linear combination of Hurwitz zeta functions.
+
+Note that this is not the same as `LSeries Φ`: they agree in the convergence range, but
+`LSeries Φ s` is defined to be `0` if `re s ≤ 1`.
+ -/
+noncomputable def LFunction (Φ : ZMod N → ℂ) (s : ℂ) : ℂ :=
+  N ^ (-s) * ∑ j : ZMod N, Φ j * hurwitzZeta (toAddCircle j) s
+
+/-- The L-function of a function on `ZMod 1` is a scalar multiple of the Riemann zeta function. -/
+lemma LFunction_mod_one (Φ : ZMod 1 → ℂ) (s : ℂ) :
+    LFunction Φ s = Φ 1 * riemannZeta s := by
+  simp only [LFunction, Nat.cast_one, one_cpow, univ_unique, sum_singleton, one_mul]
+  change Φ 1 * hurwitzZeta (toAddCircle 0) s = _
+  rw [map_zero, hurwitzZeta_zero]
+
+open scoped LSeries.notation in
+/-- For `1 < re s` the congruence L-function agrees with the sum of the Dirichlet series. -/
+lemma LFunction_eq_LSeries (Φ : ZMod N → ℂ) {s : ℂ} (hs : 1 < re s) :
+    LFunction Φ s = LSeries ↗Φ s := by
+  rw [LFunction, LSeries, mul_sum, Nat.sumByResidueClasses (LSeriesSummable_of_one_lt_re Φ hs) N]
+  congr 1 with j
+  have ha : (j.val / N : ℝ) ∈ Set.Icc 0 1 := ⟨by positivity, by
+    rw [div_le_one (Nat.cast_pos.mpr <| NeZero.pos _), Nat.cast_le]
+    exact (val_lt _).le⟩
+  rw [toAddCircle_apply, ← (hasSum_hurwitzZeta_of_one_lt_re ha hs).tsum_eq, ← mul_assoc,
+    ← tsum_mul_left]
+  congr 1 with m
+  have aux0 : (m : ℂ) + ↑(j.val / N : ℝ) = ↑((j.val + N * m) / N : ℝ) := by
+    push_cast
+    rw [add_div, mul_div_cancel_left₀ _ (NeZero.ne _), add_comm]
+  have aux1 : (0 : ℝ) ≤ j.val + N * m := by positivity
+  have aux2 : (0 : ℝ) ≤ (↑N)⁻¹ := by positivity
+  have aux3 : arg (N : ℂ) ≠ π := by simpa only [natCast_arg] using Real.pi_pos.ne
+  have aux4 : ((N : ℂ) ^ s)⁻¹ ≠ 0 := by
+    simp only [ne_eq, inv_eq_zero, cpow_eq_zero_iff, NeZero.ne, false_and, not_false_eq_true]
+  rw [aux0, div_eq_mul_inv _ (N : ℝ), ofReal_mul, mul_cpow_ofReal_nonneg aux1 aux2, ← div_div,
+    ofReal_inv, ofReal_natCast, cpow_neg, inv_cpow _ _ aux3, ← mul_div_assoc, mul_assoc,
+    mul_div_cancel_left₀ _ aux4, mul_one_div, ← Nat.cast_mul, ← Nat.cast_add, ofReal_natCast,
+    LSeries.term_of_ne_zero' (ne_zero_of_one_lt_re hs), Nat.cast_add (R := ZMod _), Nat.cast_mul,
+    CharP.cast_eq_zero (R := ZMod N) (p := N), zero_mul, add_zero]
```

Comment:
Could you please consider writing this as a `calc`-mode proof? I think it would improve readability, and probably also maintainability of the proof.